File: backend/app/schemas/agriculture.py
"""
Pydantic schemas for Agriculture module API endpoints.
Two-stage stochastic model.
Prices and plants_per_acre are now per-crop.
"""

from pydantic import BaseModel, Field, field_validator
from typing import List, Dict
from .common import Scenario


class FarmResources(BaseModel):
    """Farm-level constraints."""
    total_land: float = Field(..., gt=0, description="Total available land in acres")


class Crop(BaseModel):
    """Individual crop (berry variety) configuration."""
    name: str = Field(..., description="Crop name (e.g., 'Mohawk')")
    base_yield_per_plant: float = Field(..., gt=0, description="Base yield in lbs per plant")
    plants_per_acre: float = Field(..., gt=0, description="Number of plants per acre for this variety")
    materials_cost_per_acre: float = Field(..., ge=0, description="Materials/planting cost per acre ($)")
    purchase_price_per_lb: float = Field(..., ge=0, description="Cost to purchase from local vendor ($/lb)")
    requirement: float = Field(..., ge=0, description="Required delivery in lbs")
    selling_price: float = Field(..., ge=0, description="Selling price to buyer ($/lb)")
    selling_price_own_store: float = Field(..., ge=0, description="Selling price at own store for surplus ($/lb)")


class AgricultureScenario(Scenario):
    """Weather scenario for agriculture."""
    yield_changes: Dict[str, float] = Field(
        ...,
        description="Yield change fractions per crop (e.g., {'Mohawk': 0.1} for +10%)"
    )


class AgricultureInput(BaseModel):
    """Complete input for agriculture optimization."""
    farm_resources: FarmResources
    crops: List[Crop] = Field(..., min_length=1, description="List of berry varieties to optimize")
    scenarios: List[AgricultureScenario] = Field(..., min_length=1, description="Weather scenarios")

    @field_validator('scenarios')
    @classmethod
    def validate_scenario_probabilities(cls, v):
        total_prob = sum(s.probability for s in v)
        if not (0.999 <= total_prob <= 1.001):
            raise ValueError(f'Scenario probabilities must sum to 1.0, got {total_prob}')
        return v

File: frontend/src/types/agriculture.ts
/**
 * TypeScript type definitions for Agriculture Module
 * Prices and plants_per_acre are per-crop.
 */

export interface FarmResources {
  total_land: number;
}

export interface Crop {
  name: string;
  base_yield_per_plant: number;      // lbs per plant
  plants_per_acre: number;           // plants per acre (crop-specific)
  materials_cost_per_acre: number;   // $/acre
  purchase_price_per_lb: number;     // $/lb from local vendor
  requirement: number;               // lbs required to deliver
  selling_price: number;             // $/lb to buyer
  selling_price_own_store: number;   // $/lb at own store (surplus)
}

export interface AgricultureScenario {
  name: string;
  probability: number;
  yield_changes: Record<string, number>; // fraction: 0.1 = +10%
}

export interface AgricultureInput {
  farm_resources: FarmResources;
  crops: Crop[];
  scenarios: AgricultureScenario[];
}

// --- Result types ---

export interface ScenarioResult {
  scenario_name: string;
  probability: number;
  yield_realized: Record<string, number>;
  vendor_purchases: Record<string, number>;
  own_store_sales: Record<string, number>;
  own_store_revenue: number;
  vendor_cost: number;
  profit: number;
}

export interface AgricultureResult {
  status: string;
  objective_value: number;
  stage1_decisions: {
    planting_plan: Record<string, number>;
    total_land_used: number;
    land_utilization_pct: number;
  };
  stage2_decisions: ScenarioResult[];
  financial_summary: {
    buyer_revenue: number;
    planting_cost: number;
    expected_profit: number;
    worst_case_profit: number;
    best_case_profit: number;
    profit_range: number;
  };
  risk_metrics: {
    expected_profit: number;
    std_deviation: number;
    min_profit: number;
    max_profit: number;
    profit_range: number;
  };
}
